treq/auth.py

```diff
@@ -26,6 +182,112 @@ def request(self, method, uri, headers=None, bodyProducer=None):
             method, uri, headers=headers, bodyProducer=bodyProducer)
 
 
+class _RequestDigestAuthenticationAgent(object):
+
+    digest_auth_cache = {}
```

Comment:
treq tries to avoid having global mutable state like this cache.

The reason for this is that with a global cache like this it is very easy to implement it in such a way that different execution contexts can clobber the cached data.

For example, let's say we have a service that posts payloads to user configured URLs with digest authentication.

If two different users of this service configure it to post to same URI with different credentials they 'https://foo.bar/baz' they will end up sharing the cache entry for `('POST', 'https://foo.bar/baz')` and one of them would never be able to authenticate.

Rather than add the credentials to cache key though it'd be better if the caller of the treq request methods had to opt into the cache. 

 The only other place where there is some expectation of the top-level treq request functions would reuse state is for cookie storage where the user has to explicitly pass a CookieJar instance (instead of a dict) to have it reused. Perhaps this cache should be on the credentials object instead then?

Such that

```python
# Has one cache
treq.get(someURL, auth=HTTPDigestAuth('foo', 'bar'))

# Has a different cache
treq.get(someURL, auth=HTTPDigestAuth('baz', 'bax'))

# and share a cache.
creds = HTTPDigestAuth('foo', 'bar')
treq.get(someURL, auth=creds)
treq.get(someURL, auth=creds)

it would also make sense to augment `treq.client.HTTPClient` to take auth as a parameter to __init__ such that:

```python

client = treq.client.HTTPClient(baseAgent, auth=HTTPDigestAuth('foo', 'bar'))

# These would share auth credentials and a cache.
client.get(someURL)
client.get(someURL)
```